### ISSUE TITLE
Consider mozilla/sccache instead of Swatinem/rust-cache

### ISSUE THREAD

**Issue by galargh** [libp2p/rust-libp2p#3823]
With the majority of build jobs moving to self-hosted runners, it might be a good idea to consider using [sccache on S3](https://github.com/mozilla/sccache/blob/main/docs/S3.md) instead of [Swatinem/rust-cache](https://github.com/Swatinem/rust-cache).

The sccache can grow quite quickly which makes it less than ideal candidate as a default cache for hosted GHA runners. With self-hosted the cache size should less of an issue so it might provide further speed-ups.

As a fallback, sccache can be also configured to work with [hosted GHA runners](https://github.com/mozilla/sccache/blob/main/docs/GHA.md).

**Comment by thomaseizinger** [comment]
Is the download of artifacts from S3 on-demand with sccache? Or is it going to download everything at the start?

The former would give us a nice speed up I think because at the moment, we are storing several things in our caches that might not get used by a job (i.e. artifacts from different MSRV, etc).

**Comment by thomaseizinger** [comment]
I think this could significantly simplify our caching setup.

At the moment, we have a dedicated cache factory workflow that warms up caches on each merge into master. But those are big because they need to serve various different MSRVs. I think life would be a lot simpler if we would just plug sccache with an S3 backend in front of every compilation, including testing, clippy etc.

What is important is that is also works for our contributors, i.e. when PRs come from forks.

I don't have any bandwidth for this unfortunately :(

**Comment by galargh** [comment]
> Is the download of artifacts from S3 on-demand with sccache? Or is it going to download everything at the start?

I don't think so. I'm pretty sure it's download everything at once. But maybe we could set up our own proxy to act like that? We're already doing a similar thing for Go with Go modules proxy running on ECS and caching artifacts in S3. How does downloading deps work in Rust? Is it just getting stuff from crates.io or are other places involved too?

**Comment by thomaseizinger** [comment]
> > Is the download of artifacts from S3 on-demand with sccache? Or is it going to download everything at the start?
> 
> I don't think so. I'm pretty sure it's download everything at once. But maybe we could set up our own proxy to act like that? We're already doing a similar thing for Go with Go modules proxy running on ECS and caching artifacts in S3. How does downloading deps work in Rust? Is it just getting stuff from crates.io or are other places involved too?

Is there some form of GC happening? With changing versions of dependencies, how do we stop this cache from growing forever?

Implementing a custom proxy sounds like a lot of work, I'd like to avoid it :)

It is just crates.io for us. But it is just source code, everything is compiled locally and build-script shipped with crates often make this machine-specific. I'd expect sccache to handle all of that though.

**Comment by galargh** [comment]
> Is there some form of GC happening? With changing versions of dependencies, how do we stop this cache from growing forever?

There are lifecycle rules configured.

**Comment by thomaseizinger** [comment]
> > Is there some form of GC happening? With changing versions of dependencies, how do we stop this cache from growing forever?
> 
> There are lifecycle rules configured.

In that case, I don't think it is an issue if we download the entire cache at once. Do you have capacity to help us set this up?
